### ISSUE TITLE
[Ecosystem] UNIDQ

### ISSUE THREAD

**Issue by Shivakoreddi** [pytorch-fdn/ecosystem#51]
### Contact emails

shivacse14@gmail.com, sravani.sowrupilli@gmail.com,sowrupilli.sravani@gmail.com

### Project summary

A unified transformer architecture for multi-task data quality assessment. UNIDQ handles error detection, data repair, imputation, label noise detection, and data valuation with a single model.

### Project description

UNIDQ (Unified Data Quality) is a deep learning library built on PyTorch that handles multiple data quality tasks with a single transformer-based model. Instead of using separate tools for each task, UNIDQ provides one unified solution for:

• Error Detection - Find wrong/dirty values in datasets
• Data Repair - Automatically fix detected errors
• Missing Value Imputation - Fill in missing values intelligently
• Label Noise Detection - Identify mislabeled samples
• Label Classification - Predict labels for data
• Data Valuation - Score each sample's quality

Key Features:
- Unified architecture handling 6 tasks simultaneously
- Fast: Processes 200K records in under 20 minutes
- Simple and intuitive API
- State-of-the-art accuracy on benchmark datasets
- Efficient transformer-based architecture
- Seamless integration with Pandas DataFrames
- Pre-trained model support

UNIDQ achieves strong performance: F1=0.89 for error detection, R²=0.94 for imputation, and F1=0.86 for label noise detection.

### Are there any other projects in the PyTorch Ecosystem similar to yours?  If, yes, what are they?

cleanlab, Datawig, PyOD - but UNIDQ is unique in multi-task unified approach

### Project repo URL

https://github.com/Shivakoreddi/unidq

### Additional repos in scope of the application

NA - All code is in the primary repository

### Project license

MIT

### GitHub handles of the project maintainer(s)

Shivakoreddi, sravanisowrupilli

### Is there a corporate or academic entity backing this project?  If so, please provide the name and URL of the entity.

No corporate or academic backing at this time. This is an independent open-source project.

### Website URL

https://pypi.org/project/unidq/

### Documentation

https://github.com/Shivakoreddi/unidq#readme

### How do you build and test the project today (continuous integration)?  Please describe.

**Build Process:**
- Standard Python packaging using `setuptools` and `build`
- Automated builds via GitHub Actions (planned)
- PyPI deployment using `twine`

**Testing:**
- Unit tests using `pytest` for all core components:
  - Model architecture (test_model.py)
  - Dataset handling (test_dataset.py)
  - Multi-task training and evaluation
- Test coverage includes:
  - Model initialization and forward pass
  - Save/load functionality
  - Multi-task dataset creation
  - Task-specific evaluation metrics

**Continuous Integration:**
- Manual testing currently in place
- GitHub Actions CI/CD pipeline in development
- Automated testing on Python 3.8, 3.9, 3.10, 3.11
- PyTorch compatibility testing across versions 1.9+

**Commands:**
```bash
pytest tests/ -v
python -m build
twine check dist/*

### Version of PyTorch

PyTorch >= 1.9.0

### Components of PyTorch


UNIDQ leverages the following PyTorch components:

**Core Components:**
- `torch.nn.Module` - Base model architecture
- `torch.nn.Transformer` / `torch.nn.TransformerEncoder` - Transformer-based encoder for multi-task learning
- `torch.nn.Embedding` - Token and position embeddings
- `torch.nn.Linear` - Task-specific output heads

**Training Components:**
- `torch.optim.Adam` - Default optimizer
- `torch.nn.CrossEntropyLoss` - Classification tasks (error detection, duplicate detection, outlier detection)
- `torch.nn.MSELoss` - Regression tasks (imputation, schema matching)
- `torch.utils.data.Dataset` - Custom multi-task dataset
- `torch.utils.data.DataLoader` - Batching and data loading

**Additional:**
- Automatic mixed precision training support (planned)
- GPU acceleration via CUDA
- Model save/load using `torch.save` and `torch.load`
- Distributed training capabilities (planned)

**Why PyTorch:**
PyTorch's flexibility and dynamic computation graphs make it ideal for multi-task learning where different tasks may require different loss functions and output heads.

### How long do you expect to maintain the project?

**Long-term commitment (3+ years minimum)**

We are committed to maintaining and actively developing UNIDQ for the foreseeable future:

**Maintenance Plan:**
- **Bug fixes and updates:** Ongoing, with priority on critical issues
- **PyTorch compatibility:** Will maintain compatibility with new PyTorch releases
- **Feature development:** Roadmap includes:
  - Pre-trained model weights (Q1 2026)
  - Additional data quality tasks (Q2 2026)
  - Performance optimizations (ongoing)
  - Distributed training support (Q3 2026)
  - Integration with popular data frameworks (ongoing)

**Community Support:**
- Active response to GitHub issues
- Regular releases with changelogs
- Community contributions welcomed and encouraged

**Sustainability:**
- Growing user base with PyPI downloads
- Active development with frequent commits
- Clear documentation and examples for contributors
- Potential academic research publications to drive adoption

This is not a prototype or experiment—it's a production-ready tool we use ourselves and plan to continuously improve.

### Additional information

**Why UNIDQ Belongs in PyTorch Ecosystem:**

1. **Built for PyTorch Users:** UNIDQ is designed specifically for PyTorch users who need to clean and validate their datasets before training models. It integrates seamlessly with existing PyTorch workflows.

2. **Fills a Critical Gap:** Data quality is often the biggest bottleneck in ML pipelines, yet there's no comprehensive PyTorch-native solution. UNIDQ fills this gap.

3. **Research Foundation:** Based on cutting-edge research in multi-task learning and transformer architectures for tabular data, with planned academic publication.

4. **Production Ready:** 
   - Comprehensive test suite
   - Clear documentation and examples
   - Published on PyPI
   - Active GitHub repository

5. **Community Benefit:**
   - Simple API lowers barrier to entry for data quality tasks
   - Unified approach reduces complexity in ML pipelines
   - Educational value for students learning multi-task learning

6. **Future Roadmap:**
   - Integration with PyTorch Lightning
   - TorchScript export support
   - ONNX compatibility
   - Hugging Face Hub integration for pre-trained models

**Performance Metrics:**
- PyPI: 492


We believe UNIDQ will be valuable to the PyTorch community, especially data scientists and ML engineers dealing with real-world messy data.

**Comment by lantiga** [comment]
Thank you for the contribution @Shivakoreddi 

The project looks interesting and high quality, but at its starting stages. One of the key requirements for the ecosystem is a demonstrable degree of maturity and adoption.

We'll be happy to consider the project for a spotlight on the PyTorch blog once it gets some initial adoption, and we encourage resubmission to the ecosystem a few months after that.
